File: java/L900/LeetCode0909.java
package L900;

import java.util.Deque;
import java.util.HashMap;
import java.util.LinkedList;

public class LeetCode0909 {
    public int snakesAndLadders(int[][] board) {
        final int N = board.length;
        int[] arr = new int[N * N];
        int index = 0;
        boolean flag = true;
        for (int i = N - 1;i >= 0;--i) {
            if (flag) {
                for (int j = 0;j < N;++j) {
                    arr[index] = board[i][j];
                    ++index;
                }
            } else {
                for (int j = N - 1;j >= 0;--j) {
                    arr[index] = board[i][j];
                    ++index;
                }
            }
            flag = !flag;
        }
        Deque<Integer> q1 = new LinkedList<>();
        HashMap<Integer,Integer> m1 = new HashMap<>();
        q1.offer(0);
        m1.put(0, 0);
        int ans = N * N;
        while (!q1.isEmpty()) {
            int cur = q1.poll();
            int step = m1.get(cur);
            if (cur == N * N - 1) {
                ans = Math.min(ans, step);
                continue;
            }
            for (int i = 1;i <= 6;++i) {
                int nx = cur + i;
                if (nx < N * N) {
                    if (arr[nx] != -1) {
                        nx = arr[nx] - 1;
                    } 
                    if (!m1.containsKey(nx)) {
                        m1.put(nx, step + 1);
                        q1.offer(nx);
                    }
                }
            }
        }
        return ans == N * N ? -1 : ans;
    }
}

File: cpp/L900/LeetCode0909.cc
#include <queue>
#include <unordered_map>
#include <vector>
using std::vector;

class Solution {
public:
    int snakesAndLadders(vector<vector<int>>& board) {
        int N = board.size();
        std::unordered_map<int, int> mp;
        for (int x = N - 1, y = 0, num{1}, dy{1};num <= N * N;++num) {
            mp[num] = board[x][y];
            y += dy;
            if (y == N || y == -1) {
                y -= dy;
                x -= 1;
                dy = -dy;
            }
        }
        std::queue<int> q;
        q.push(1);
        int step{0};
        vector<bool> visited(N * N + 1, false);
        while (!q.empty()) {
            int sz = q.size();
            for (int s = 0;s < sz;++s) {
                int cur{q.front()};
                q.pop();
                visited[cur] = true;
                if (cur == N * N) {
                    return step;
                }
                for (int i = 1;i <= 6;++i) {
                    int next = std::min(cur + i, N * N);
                    if (mp[next] != -1) {
                        next = mp[next];
                    }
                    if (!visited[next]) {
                        visited[next] = true;
                        q.push(next);
                    }
                }
            }
            ++step;
        }
        return -1;
    }
};

int main() {
    vector<vector<int>> v{{-1,-1,-1,-1,-1,-1},{-1,-1,-1,-1,-1,-1},{-1,-1,-1,-1,-1,-1},{-1,35,-1,-1,13,-1},{-1,-1,-1,-1,-1,-1},{-1,15,-1,-1,-1,-1}};
    Solution s;
    s.snakesAndLadders(v);
}